app/hooks/files.ts

```diff
@@ -36,3 +51,125 @@ export const useImageAttachments = (filesInfo: FileInfo[], publicLinkEnabled: bo
     }, [filesInfo, publicLinkEnabled]);
 };
 
+export const useDownloadFileAndPreview = () => {
+    const serverUrl = useServerUrl();
+    const intl = useIntl();
+    const theme = useTheme();
+    const downloadTask = useRef<ProgressPromise<ClientResponse>>();
+
+    const [progress, setProgress] = useState<number>(0);
+
+    const [preview, setPreview] = useState(false);
+    const [downloading, setDownloading] = useState(false);
+    const [didCancel, setDidCancel] = useState(false);
+
+    let client: Client | undefined;
+    try {
+        client = NetworkManager.getClient(serverUrl);
+    } catch {
+        // do nothing
+    }
+
+    const openFile = (file: FileInfo) => {
```

Comment:
It looks pretty good. Now let's make the return of the hook "stable". That is, two calls to the hook will return the same values.

In order to do so, you will need to use `useCallback` for `toggleDownloadAndPreview`, which will lead you down to use it in many other places and reorder the code.

Reply:
Done! Can you review it?

I found it strange that the Lint of the Project didn't signal which dependencies to use.